### PR TITLE
Add  API to observe registration life cycle at client side

### PR DESCRIPTION
**Pull request by sbernard31**

### CODE REVIEW

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/observer/LwM2mClientObserver.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    void updateFailed(DmServerInfo server, ResponseCode responseCode, String errorMessage);
+
+    void updateTimeout(DmServerInfo server);
+
+    void deregisterSucceed(DmServerInfo server, String registrationID);
+
+    void deregisterFailed(DmServerInfo server, ResponseCode responseCode, String errorMessage);
+
+    void deregisterTimeout(DmServerInfo server);
+}
```
Comment (sophokles73) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/observer/LwM2mClientObserver.java]: Most handlers use a naming pattern for methods following the `onSomeEvent` pattern. What do you think about using this here as well, e.g. `onBootstrapSucceeded` and `onRegistrationFailed` etc?

File: leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/LeshanClient.java [hunk 2]
```diff
@@ -98,18 +103,33 @@ public LeshanClient(final String endpoint, final InetSocketAddress localAddress,
         // Create sender
         requestSender = new CaliforniumLwM2mClientRequestSender(secureEndpoint, nonSecureEndpoint);
 
-        // Create registration engine
-        bootstrapHandler = new BootstrapHandler(this.objectEnablers);
-        bootstrapHandler.addBootstrapListener(new BootstrapListener() {
```
Comment (msangoi) [inline leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/LeshanClient.java]: The BootstrapListener is probably not used anymore. We should remove it and rely on the LwM2mClientObserver events.

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/observer/LwM2mClientObserver.java (hunk 1 above)
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/observer/LwM2mClientObserver.java]: I will do that.

File: leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/LeshanClient.java (hunk 2 above)
Comment (sbernard31) [inline leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/LeshanClient.java]: I will do that too.

**Pull request closed without merging**